### ISSUE TITLE
SpecCluster + Dask JobQueue does not retire workers when scaling down.

### ISSUE THREAD

**Issue by bbockelm** [dask/distributed#4069]
I am currently chasing what appears to be a bug in the combination of `distributed` and `dask-jobqueue`.

Particularly, I find that this piece of code almost never completes when running a dask-jobqueue `HTCondorCluster` with adaptive scaling enabled:
```
import time
def square(x):
    time.sleep(0.5)
    return x ** 2

def neg(x):
    return -x

A = client.map(square, range(1000))
B = client.map(neg, A)
total = client.submit(sum, B)
total.result()
```

What happens is that, whenever the tasks get almost finished, one of the workers is killed.  This results in lost keys, sets the progress back a few hundred tasks, and repeats indefinitely.

Putting some debugging in the decision making, it appears that the adaptive code is deciding to scale down whenever it decides there is insufficient work to support all the workers.  This results in the invocation of the following code in `SpecCluster`:

https://github.com/dask/distributed/blob/2.24.0/distributed/deploy/spec.py#L327-L329

Particularly, the list `to_close` comes from `self.workers` and appears to be integers (e.g., `0`) and invokes the `scheduler.retire_workers` method.  Adding a print statement, I see the following worker dictionary:

```
{0: <CoffeaCasaJob: status=Status.running>, 1: <CoffeaCasaJob: status=Status.running>,
 2: <CoffeaCasaJob: status=Status.running>, 3: <CoffeaCasaJob: status=Status.running>,
 4: <CoffeaCasaJob: status=Status.running>, 5: <CoffeaCasaJob: status=Status.running>}
```

However, that method filters passed workers on a different set of keys:

https://github.com/dask/distributed/blob/93a8d28933ebe4b82d715e9f6948be3c38c49953/distributed/scheduler.py#L3502

Examples from adding a print statement:

```
SortedDict({'tls://129.93.X.Y:32806': <Worker 'tls://129.93.X.Y:32806', name: htcondor--4523817.0--,
 memory: 223, processing: 46>, 'tls://129.93.X.Y:32883': <Worker 'tls://129.93.X.Y:32883', name: htcondor--4523815.0--,
 memory: 214, processing: 47>, 'tls://129.93.X.Y:32768': <Worker 'tls://129.93.X.Y:32768', name: htcondor--4523813.0--,
 memory: 740, processing: 33>, 'tls://129.93.X.Y:32769': <Worker 'tls://129.93.X.Y:32769', name: htcondor--4523814.0--,
 memory: 204, processing: 46>, 'tls://129.93.X.Y:32815': <Worker 'tls://129.93.X.Y:32815', name: htcondor--4523816.0--,
 memory: 219, processing: 36>})
```

(IP addresses lightly redacted)

Because, for example, the key `0` does not match `tls://129.93.182.193:32806`, the set of workers to retire becomes the empty set and nothing occurs.

Next, the selected workers are immediately killed.  Because the unique data is not replicated elsewhere, the intermediate task output is lost and has to be re-run.  If you time it right, this will get into an loop until a given task has been killed enough times and becomes terminal because of a KilledWorker exception.

I don't see any immediately obvious solution as there's no standard naming scheme between the worker name and the job ID.  However, I'm not overly familiar with the code.

**Environment**:

- Dask version: 0.240.
- Python version: 3.8
- Operating System: Linux
- Install method (conda, pip, source): conda

**Comment by jrbourbeau** [comment]
Thanks for raising an issue and providing a detailed description of the situation @bbockelm! After briefly looking through the relevant code sections (thanks for the links 🙂 ), I agree there does appear to be a mismatch with how we're keeping track of workers in the `SpecCluster` and the scheduler.

It looks like we make sure a name is always specified when creating new `SpecCluster` workers:

https://github.com/dask/distributed/blob/a848ee0c20d4b3df261da562c4c66cb762d2a3ec/distributed/deploy/spec.py#L343-L345

Updating the `retire_workers` call to retire workers by their name instead of their address

https://github.com/dask/distributed/blob/93a8d28933ebe4b82d715e9f6948be3c38c49953/distributed/scheduler.py#L3479-L3487

might resolve this issue. Is this something you're interested in looking into further? (no obligation though)

**Comment by bbockelm** [comment]
Yeah, I think that ought to work.  However, it appears that we're annoyingly overriding the `--name` flag in our test setup (meaning I'm not in a good place to test out a fix).  Might need a few days to unwind why that's being done.

_That said_, I can confirm matching based on names works (if I hack up the name scheme to match our generated ones).  Just can't write a generic patch right now.

Once the names match, I see workers being appropriately retired, unique keys copied off to other locations, and scale-down without any problematic side-efffects.

**Comment by oshadura** [comment]
@bbockelm I see sometimes (very rarely) with autoscaling coffea-casa is failing: https://gist.github.com/oshadura/7dd44b495345edb1b3622117595ad047

**Comment by oshadura** [comment]
@jrbourbeau, I saw there was opened https://github.com/dask/distributed/pull/4074, any help needed there? We at coffea-casa (cc: @bbockelm @nsmith-) are using various patches/solutions to resolve this issue but it will be nice to have this solution resolved in master...
